_rules/active-focus-shortcut-1e9941.md

```diff
@@ -0,0 +1,212 @@
+- [G127 Providing a mechanism to allow users to remap or turn off character key shortcuts](https://www.w3.org/WAI/WCAG21/Techniques/general/G217)
+- [F99 Failure of Success Criterion 2.1.4 due to implementing character key shortcuts that cannot be turned off or remapped](https://www.w3.org/WAI/WCAG21/Techniques/failures/F99)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+The [HTML document][] has a [keyboard shortcut][keyboard shortcuts] that uses one [printable][printable characters] and one [non-printable characters][].
+
+```html
+<html>
+  <head>
+    <title>Passed Example 1</title>
+    <script>
+      function shortcut() {
+        document.body.addEventListener('keydown', function(event) {
+          if (event.key === 'i' && event.ctrlKey) {
+            const text = document.getElementById('text');
+            text.className = text.className === 'italic' ? '' : 'italic';
+          }
+        });
+      }
+    </script>
+    <style>
+      .italic {
+        font-style: italic;
+      }
+    </style>
+  </head>
+  <body onload="shortcut();">
+    <div>Press <strong>ctrl+i</strong> to toggle italic format</div>
+    <div id="text">Some text inside the document content</div>
+  </body>
+</html>
+```
+
+#### Passed Example 2
+
+The [HTML document][] has an element with the attribute `accesskey`. Accesskeys use [non-printable characters][].
+
+```html
+<html>
+  <head>
+    <title>Passed Example 2</title>
+    <script>
+      function shortcut() {
+        const button = document.getElementById('italic');
+
+        button.addEventListener('click', function() {
+          const text = document.getElementById('text');
+          text.className = text.className === 'italic' ? '' : 'italic';
+        });
+
+        button.textContent +=
+          button.accessKeyLabel ? ' (' + button.accessKeyLabel + ')' : ' (accesskey +' + button.accessKey + ')';
+      }
+    </script>
+    <style>
+      .italic {
+        font-style: italic;
+      }
+    </style>
+  </head>
+  <body onload="shortcut();">
+    <button id="italic" type="button" accessKey="i">Toggle italic format</button>
+    <div id="text">Some text inside the document content</div>
+  </body>
+</html>
+```
+
+#### Passed Example 3
+
+The [HTML document][] has a single [printable character][printable characters] [keyboard shortcut][keyboard shortcuts] for a [user interface component][], which is only available when that component has focus.
+
+```html
+<html>
+  <head>
+    <title>Passed Example 3</title>
+    <script>
+      function shortcut() {
+        document.body.addEventListener('keydown', function(event) {
+          const text = document.getElementById('text');
+
+          if (event.key === '+' && document.activeElement === text) {
+            document.getElementById('list').innerHTML += '<li>' + text + '</li>';
+            text.value = '';
+            event.preventDefault();
+          }
+        });
+      }
+    </script>

```

Comment:
I think this can be refactored with the code of Failed Example 1 by:
- having the anonymous function be a named function taking an `options` object parameter (and also the `text` element) and returning a function.
- having the currently named `shortcut` function renamed to `activateShortcut` which figure out what is `text`, adds the event listener passing the `options` object.
- call `onload="activateShortcut( {needFocus : true/false} )"` in the HTML.

Augmenting the `options` object (and adding a couple of tests to `activateShortcut`) should allow to use the same code for the `ctrl+key` and access key examples.

```javascript
function addItem(list, text) { /* change innerHTML and reset text value */ }

function shortcut(text, options) {
  function foo(event) {
    // test on options fields + event and call addItem if needed
  }
  return foo;
}

function activateShortcut(options) {
  // find text, add event listener passing down the options
  // check options.useButton (or similar) to also add event listener to the button (Passed Example 2).
}
```
This code should be identical (I hope…) for all your examples here (by using the same "add to list" example) and thus can be refactored in a test asset.

Reply:
Honest question: do you really think that this will improve the legibility of the examples?

Reply:
Yes, I do 😄 
The idea is that it will move most of the complexity to a separate file. And keep the differences between examples to a single list of arguments to the function, which with good names will make purpose obvious.

Typically, a Passed Example will have `onload="activateShortcut({ needFocus: true })"` while the Failed Example will have `onload="activateShortcut({ needFocus: false })"` and all the rest is identical.

Even without reading the Javascript, I can see that both these examples have a shortcut and that one of them is only active when something has focus somewhere. Within the context of the rule, it is fairly obvious what this mean.

---
Then (or before), of course, I can check the Javascript and check that the options names have the effect that they claim to have. But reading and understanding the Javascript is now an isolated exercice where I don't need to really focus on the subtleties of the rule. I can read the Javascript and understand that `activateShortcut` will indeed activate a shortcut and that if the `needFocus` parameter is set to `true`, the shortcut will only work if the component has focus. Once I'm convinced that the Javascript is doing what it claims, I can forget the details and go back to the rule, now when I see `activateShortcut({ needFocus: true })`, I am convinced that this activate a shortcut that will only be active when the component has focus.

So, yes, I do think that this will improve readability for humans 😄 

I tried to do that refactoring for @joao-vicente's device motion rule: https://drive.google.com/open?id=1gOR9_OGqcHJ8H-R4yRdNg-zGxVe56q0Z
I do feel the result is more understandable than if the code was copied in all examples.